Repository cache invalidation from method calls when discarding working directory changes

On Appveyor 

Repository cache invalidation from method calls when discarding working directory changes
repository.test.js 1 sec 278 ms

Error message:
```
bzzzt, inadequate cache busting detected: expected false to be true
```

Error stack trace:
```
AssertionError: bzzzt, inadequate cache busting detected: expected false to be true
    at doAsserterAsyncAndAddThen (C:\projects\github\node_modules\chai-as-promised\lib\chai-as-promised.js:296:33)
    at .<anonymous> (C:\projects\github\node_modules\chai-as-promised\lib\chai-as-promised.js:286:25)
    at get (C:\projects\github\node_modules\chai\lib\chai\utils\overwriteProperty.js:50:37)
    at Function.assert.isTrue (C:\projects\github\node_modules\chai\lib\chai\interface\assert.js:332:31)
    at C:/projects/github/repository.test.js:1120:16
    at next (<anonymous>)
    at step (C:/projects/github/repository.test.js:18:1)
    at C:/projects/github/repository.test.js:18:1
```

Std OUT:
```
These cached method results should have been invalidated, but were not:
 getStagedChangesSinceParentCommit:
  before: Promise {
  [ { filePath: 'b.txt', status: 'modified' },
  { filePath: 'c.txt', status: 'modified' },
  { filePath: 'subdir-1\\b.txt', status: 'modified' },
  { filePath: 'subdir-1\\c.txt', status: 'modified' } ] }
  cached: Promise {
  [ { filePath: 'b.txt', status: 'modified' },
  { filePath: 'c.txt', status: 'modified' },
  { filePath: 'subdir-1\\b.txt', status: 'modified' },
  { filePath: 'subdir-1\\c.txt', status: 'modified' } ] }
   after: Promise {
  <rejected> { Error
    at new GitError (C:/projects/github/git-shell-out-strategy.js:32:18)
    at C:/projects/github/git-shell-out-strategy.js:231:23
    at next (<anonymous>)
    at step (C:/projects/github/git-shell-out-strategy.js:19:1)
    at C:/projects/github/git-shell-out-strategy.js:19:1
  code: 128,
  stdErr: 'fatal: .git/index: index file open failed: Permission denied\n',
  stdOut: '',
  command: 'git diff --name-status --no-renames --staged HEAD~ in C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\git-fixture-11774-976-15e7iga.8i76s3v7vi' } }
```

--- a/lib/models/repository-states/present.js
+++ b/lib/models/repository-states/present.js
@@ -70,7 +70,6 @@ export default class Present extends State {
 
   acceptInvalidation(spec, args) {
     const keys = spec(...args);
-    console.log(`invalidating: ${keys.join(', ')}`);
     this.cache.invalidate(keys);
     this.didUpdate();
   }

--- a/lib/models/repository-states/present.js
+++ b/lib/models/repository-states/present.js
@@ -70,6 +70,7 @@ export default class Present extends State {
 
   acceptInvalidation(spec, args) {
     const keys = spec(...args);
+    console.log(`invalidating: ${keys.join(', ')}`);
     this.cache.invalidate(keys);
     this.didUpdate();
   }
@@ -754,6 +755,10 @@ class CacheKey {
       cache.removeFromGroup(group, this);
     }
   }
+
+  toString() {
+    return `CacheKey(${this.primary})`;
+  }
 }
 
 class GroupKey {
@@ -765,6 +770,10 @@ class GroupKey {
     for (const matchingKey of cache.keysInGroup(this.group)) {
       matchingKey.removeFromCache(cache, this.group);
     }
+  }
+
+  toString() {
+    return `GroupKey(${this.group})`;
   }
 }
 

--- a/lib/models/repository-states/state.js
+++ b/lib/models/repository-states/state.js
@@ -353,6 +353,11 @@ export default class State {
     return Promise.resolve(nullBranch);
   }
 
+  @shouldDelegate
+  getHeadDescription() {
+    return Promise.resolve('(no repository)');
+  }
+
   // Merging and rebasing status
 
   @shouldDelegate

--- a/lib/models/repository.js
+++ b/lib/models/repository.js
@@ -288,6 +288,7 @@ const delegates = [
 
   'getBranches',
   'getCurrentBranch',
+  'getHeadDescription',
 
   'isMerging',
   'isRebasing',

--- a/test/models/repository.test.js
+++ b/test/models/repository.test.js
@@ -930,7 +930,10 @@ describe('Repository', function() {
     });
   });
 
-  describe('cache invalidation', function() {
+  describe.only('cache invalidation', function() {
+    // These tests do a *lot* of git operations
+    this.timeout(this.timeout() * 2);
+
     const preventDefault = event => event.preventDefault();
 
     beforeEach(function() {
@@ -1023,12 +1026,12 @@ describe('Repository', function() {
         withFile(fileName);
       }
 
-      const withBranch = (branchName, description) => {
-        calls.set(`getAheadCount ${description}`, {
+      const withBranch = branchName => {
+        calls.set(`getAheadCount ${branchName}`, {
           serial: true,
           op: () => repository.getAheadCount(branchName),
         });
-        calls.set(`getBehindCount ${description}`, {
+        calls.set(`getBehindCount ${branchName}`, {
           serial: true,
           op: () => repository.getBehindCount(branchName),
         });
@@ -1209,7 +1212,7 @@ describe('Repository', function() {
         const patch = await repository.getFilePatchForPath('a.txt');
         await writeFile(path.join(workdir, 'a.txt'), 'foo\nfoo-1\nfoo-2\n');
 
-        await assertCorrectInvalidation({repository}, async () => {
+        await assertCorrectInvalidation({repository, verbose: true}, async () => {
           await repository.applyPatchToIndex(patch);
         });
       });

--- a/test/models/repository.test.js
+++ b/test/models/repository.test.js
@@ -1386,19 +1386,21 @@ describe('Repository', function() {
         return {repository, observer};
       }
 
-      function expectEvents(repository, ...fileNames) {
-        const pending = new Set(
-          fileNames.map(fileName => {
-            return fs.realpathSync(
-              path.join(repository.getWorkingDirectoryPath(), ...fileName.split(/[\\/]/)),
-            );
-          }),
-        );
+      function expectEvents(repository, ...suffixes) {
+        const pending = new Set(suffixes);
         return new Promise((resolve, reject) => {
           eventCallback = () => {
             const matchingPaths = observedEvents
               .map(event => fs.realpathSync(event.path))
-              .filter(eventPath => pending.delete(eventPath));
+              .filter(eventPath => {
+                for (const suffix of pending) {
+                  if (eventPath.endsWith(suffix)) {
+                    pending.delete(suffix);
+                    return true;
+                  }
+                }
+                return false;
+              });
 
             if (matchingPaths.length > 0) {
               repository.observeFilesystemChange(matchingPaths);

--- a/test/models/repository.test.js
+++ b/test/models/repository.test.js
@@ -932,7 +932,7 @@ describe('Repository', function() {
 
   describe('cache invalidation', function() {
     // These tests do a *lot* of git operations
-    this.timeout(this.timeout() * 2);
+    this.timeout(Math.max(20000, this.timeout() * 2));
 
     const preventDefault = event => event.preventDefault();
 

--- a/test/models/repository.test.js
+++ b/test/models/repository.test.js
@@ -1380,7 +1380,6 @@ describe('Repository', function() {
 
         sub.add(observer.onDidChange(events => {
           observedEvents.push(...events);
-          console.log(`> events received:\n ${events.map(e => e.path).join('\n ')}`);
           eventCallback();
         }));
 
@@ -1397,8 +1396,6 @@ describe('Repository', function() {
                 for (const suffix of pending) {
                   if (eventPath.endsWith(suffix)) {
                     pending.delete(suffix);
-                    console.log(`> event ${eventPath} matched ${suffix}`);
-                    console.log(`> remaining: ${Array.from(pending).join(', ')}`)
                     return true;
                   }
                 }
@@ -1410,13 +1407,11 @@ describe('Repository', function() {
             }
 
             if (pending.size === 0) {
-              console.log('> all events received');
               resolve();
             }
           };
 
           if (observedEvents.length > 0) {
-            console.log('> triggering initial callback');
             eventCallback();
           }
         });
@@ -1493,28 +1488,21 @@ describe('Repository', function() {
         });
       });
 
-      stress(400, 'when committing', async function() {
-        console.log('---> start');
+      it('when committing', async function() {
         const {repository, observer} = await wireUpObserver();
 
         await writeFile(path.join(workdir, 'a.txt'), 'boop\n');
         await repository.stageFiles(['a.txt']);
 
-        console.log('---> collecting baseline');
         await assertCorrectInvalidation({repository}, async () => {
-          console.log('---> starting watcher');
           await observer.start();
-          console.log('---> performing commit');
           await repository.git.commit('boop your snoot');
-          console.log('---> waiting for filesystem events');
           await expectEvents(
             repository,
             path.join('.git', 'index'),
             path.join('.git', 'refs', 'heads', 'master'),
           );
-          console.log('---> collecting cached data');
         });
-        console.log('---> complete');
       });
 
       it('when merging', async function() {

--- a/test/models/repository.test.js
+++ b/test/models/repository.test.js
@@ -1463,7 +1463,7 @@ describe('Repository', function() {
         });
       });
 
-      stress(50, 'when applying a patch to the index', async function() {
+      it('when applying a patch to the index', async function() {
         const {repository, observer} = await wireUpObserver();
 
         await writeFile(path.join(workdir, 'a.txt'), 'boop\n');

--- a/test/models/repository.test.js
+++ b/test/models/repository.test.js
@@ -1394,8 +1394,6 @@ describe('Repository', function() {
         );
 
         sub.add(observer.onDidChange(events => {
-          const paths = events.map(e => e.path);
-          repository.observeFilesystemChange(paths);
           observedEvents.push(...events);
           eventCallback();
         }));
@@ -1403,11 +1401,19 @@ describe('Repository', function() {
         return {repository, observer};
       }
 
-      function expectEvents(...fileNames) {
+      function expectEvents(repository, ...fileNames) {
+        const pending = new Set(fileNames);
         return new Promise((resolve, reject) => {
           eventCallback = () => {
-            const observedFileNames = new Set(observedEvents.map(event => path.basename(event.path)));
-            if (fileNames.every(fileName => observedFileNames.has(fileName))) {
+            const matchingPaths = observedEvents
+              .map(event => event.path)
+              .filter(eventPath => pending.delete(path.basename(eventPath)));
+
+            if (matchingPaths.length > 0) {
+              repository.observeFilesystemChange(matchingPaths);
+            }
+
+            if (pending.size === 0) {
               resolve();
             }
           };
@@ -1427,10 +1433,10 @@ describe('Repository', function() {
 
         await writeFile(path.join(workdir, 'a.txt'), 'boop\n');
 
-        await observer.start();
         await assertCorrectInvalidation({repository}, async () => {
+          await observer.start();
           await repository.git.stageFiles(['a.txt']);
-          await expectEvents('index');
+          await expectEvents(repository, 'index');
         });
       });
 
@@ -1440,33 +1446,33 @@ describe('Repository', function() {
         await writeFile(path.join(workdir, 'a.txt'), 'boop\n');
         await repository.git.stageFiles(['a.txt']);
 
-        await observer.start();
         await assertCorrectInvalidation({repository}, async () => {
+          await observer.start();
           await repository.git.unstageFiles(['a.txt']);
-          await expectEvents('index');
+          await expectEvents(repository, 'index');
         });
       });
 
       it('when staging files from a parent commit', async function() {
         const {repository, observer} = await wireUpObserver();
 
-        await observer.start();
         await assertCorrectInvalidation({repository}, async () => {
+          await observer.start();
           await repository.git.unstageFiles(['a.txt'], 'HEAD~');
-          await expectEvents('index');
+          await expectEvents(repository, 'index');
         });
       });
 
-      it('when applying a patch to the index', async function() {
+      stress(50, 'when applying a patch to the index', async function() {
         const {repository, observer} = await wireUpObserver();
 
         await writeFile(path.join(workdir, 'a.txt'), 'boop\n');
         const patch = await repository.getFilePatchForPath('a.txt');
 
-        await observer.start();
         await assertCorrectInvalidation({repository}, async () => {
+          await observer.start();
           await repository.git.applyPatch(patch.getHeaderString() + patch.toString(), {index: true});
-          await expectEvents('index');
+          await expectEvents(repository, 'index');
         });
       });
 
@@ -1476,10 +1482,10 @@ describe('Repository', function() {
         await writeFile(path.join(workdir, 'a.txt'), 'boop\n');
         const patch = (await repository.getFilePatchForPath('a.txt')).getUnstagePatch();
 
-        await observer.start();
         await assertCorrectInvalidation({repository}, async () => {
+          await observer.start();
           await repository.git.applyPatch(patch.getHeaderString() + patch.toString());
-          await expectEvents('a.txt');
+          await expectEvents(repository, 'a.txt');
         });
       });
 
@@ -1489,20 +1495,20 @@ describe('Repository', function() {
         await writeFile(path.join(workdir, 'a.txt'), 'boop\n');
         await repository.stageFiles(['a.txt']);
 
-        await observer.start();
         await assertCorrectInvalidation({repository}, async () => {
+          await observer.start();
           await repository.git.commit('boop your snoot');
-          await expectEvents('index', 'master');
+          await expectEvents(repository, 'index', 'master');
         });
       });
 
       it('when merging', async function() {
         const {repository, observer} = await wireUpObserver('merge-conflict');
 
-        await observer.start();
         await assertCorrectInvalidation({repository}, async () => {
+          await observer.start();
           await assert.isRejected(repository.git.merge('origin/branch'));
-          await expectEvents('index', 'modified-on-both-ours.txt', 'MERGE_HEAD');
+          await expectEvents(repository, 'index', 'modified-on-both-ours.txt', 'MERGE_HEAD');
         });
       });
 
@@ -1510,30 +1516,30 @@ describe('Repository', function() {
         const {repository, observer} = await wireUpObserver('merge-conflict');
         await assert.isRejected(repository.merge('origin/branch'));
 
-        await observer.start();
         await assertCorrectInvalidation({repository}, async () => {
+          await observer.start();
           await repository.git.abortMerge();
-          await expectEvents('index', 'modified-on-both-ours.txt', 'MERGE_HEAD', 'HEAD');
+          await expectEvents(repository, 'index', 'modified-on-both-ours.txt', 'MERGE_HEAD', 'HEAD');
         });
       });
 
       it('when checking out a revision', async function() {
         const {repository, observer} = await wireUpObserver();
 
-        await observer.start();
         await assertCorrectInvalidation({repository}, async () => {
+          await observer.start();
           await repository.git.checkout('HEAD^');
-          await expectEvents('index', 'HEAD', 'b.txt', 'c.txt');
+          await expectEvents(repository, 'index', 'HEAD', 'b.txt', 'c.txt');
         });
       });
 
       it('when checking out paths', async function() {
         const {repository, observer} = await wireUpObserver();
 
-        await observer.start();
         await assertCorrectInvalidation({repository}, async () => {
+          await observer.start();
           await repository.git.checkoutFiles(['b.txt'], 'HEAD^');
-          await expectEvents('b.txt', 'index');
+          await expectEvents(repository, 'b.txt', 'index');
         });
       });
 
@@ -1544,10 +1550,10 @@ describe('Repository', function() {
         await repository.commit('wat', {allowEmpty: true});
         await repository.commit('huh', {allowEmpty: true});
 
-        await observer.start();
         await assertCorrectInvalidation({repository}, async () => {
+          await observer.start();
           await repository.git.fetch('origin', 'master');
-          await expectEvents('master');
+          await expectEvents(repository, 'master');
         });
       });
 
@@ -1559,10 +1565,10 @@ describe('Repository', function() {
         await repository.stageFiles(['file.txt']);
         await repository.commit('wat');
 
-        await observer.start();
         await assertCorrectInvalidation({repository}, async () => {
+          await observer.start();
           await assert.isRejected(repository.git.pull('origin', 'master'));
-          await expectEvents('file.txt', 'master', 'MERGE_HEAD', 'index');
+          await expectEvents(repository, 'file.txt', 'master', 'MERGE_HEAD', 'index');
         });
       });
 
@@ -1574,10 +1580,10 @@ describe('Repository', function() {
         await repository.stageFiles(['new-file.txt']);
         await repository.commit('wat');
 
-        await observer.start();
         await assertCorrectInvalidation({repository}, async () => {
+          await observer.start();
           await repository.git.push('origin', 'master');
-          await expectEvents('master');
+          await expectEvents(repository, 'master');
         });
       });
 
@@ -1585,20 +1591,20 @@ describe('Repository', function() {
         const {repository, observer} = await wireUpObserver();
 
         const optionNames = ['core.editor', 'color.ui'];
-        await observer.start();
         await assertCorrectInvalidation({repository, optionNames}, async () => {
+          await observer.start();
           await repository.git.setConfig('core.editor', 'ed # :trollface:');
-          await expectEvents('config');
+          await expectEvents(repository, 'config');
         });
       });
 
       it('when changing files in the working directory', async function() {
         const {repository, observer} = await wireUpObserver();
 
-        await observer.start();
         await assertCorrectInvalidation({repository}, async () => {
+          await observer.start();
           await writeFile(path.join(workdir, 'b.txt'), 'new contents\n');
-          await expectEvents('b.txt');
+          await expectEvents(repository, 'b.txt');
         });
       });
     });

--- a/test/models/repository.test.js
+++ b/test/models/repository.test.js
@@ -1380,6 +1380,7 @@ describe('Repository', function() {
 
         sub.add(observer.onDidChange(events => {
           observedEvents.push(...events);
+          console.log(`> events received:\n ${events.map(e => e.path).join('\n ')}`);
           eventCallback();
         }));
 
@@ -1396,6 +1397,8 @@ describe('Repository', function() {
                 for (const suffix of pending) {
                   if (eventPath.endsWith(suffix)) {
                     pending.delete(suffix);
+                    console.log(`> event ${eventPath} matched ${suffix}`);
+                    console.log(`> remaining: ${Array.from(pending).join(', ')}`)
                     return true;
                   }
                 }
@@ -1407,11 +1410,13 @@ describe('Repository', function() {
             }
 
             if (pending.size === 0) {
+              console.log('> all events received');
               resolve();
             }
           };
 
           if (observedEvents.length > 0) {
+            console.log('> triggering initial callback');
             eventCallback();
           }
         });
@@ -1488,21 +1493,28 @@ describe('Repository', function() {
         });
       });
 
-      it('when committing', async function() {
+      stress(100, 'when committing', async function() {
+        console.log('---> start');
         const {repository, observer} = await wireUpObserver();
 
         await writeFile(path.join(workdir, 'a.txt'), 'boop\n');
         await repository.stageFiles(['a.txt']);
 
+        console.log('---> collecting baseline');
         await assertCorrectInvalidation({repository}, async () => {
+          console.log('---> starting watcher');
           await observer.start();
+          console.log('---> performing commit');
           await repository.git.commit('boop your snoot');
+          console.log('---> waiting for filesystem events');
           await expectEvents(
             repository,
             path.join('.git', 'index'),
             path.join('.git', 'refs', 'heads', 'master'),
           );
+          console.log('---> collecting cached data');
         });
+        console.log('---> complete');
       });
 
       it('when merging', async function() {

--- a/test/models/repository.test.js
+++ b/test/models/repository.test.js
@@ -930,7 +930,7 @@ describe('Repository', function() {
     });
   });
 
-  describe.only('cache invalidation', function() {
+  describe('cache invalidation', function() {
     // These tests do a *lot* of git operations
     this.timeout(this.timeout() * 2);
 

--- a/test/models/repository.test.js
+++ b/test/models/repository.test.js
@@ -1493,7 +1493,7 @@ describe('Repository', function() {
         });
       });
 
-      stress(100, 'when committing', async function() {
+      stress(400, 'when committing', async function() {
         console.log('---> start');
         const {repository, observer} = await wireUpObserver();
 

--- a/test/models/repository.test.js
+++ b/test/models/repository.test.js
@@ -1391,7 +1391,7 @@ describe('Repository', function() {
         return new Promise((resolve, reject) => {
           eventCallback = () => {
             const matchingPaths = observedEvents
-              .map(event => fs.realpathSync(event.path))
+              .map(event => event.path)
               .filter(eventPath => {
                 for (const suffix of pending) {
                   if (eventPath.endsWith(suffix)) {

--- a/test/models/repository.test.js
+++ b/test/models/repository.test.js
@@ -1197,7 +1197,7 @@ describe('Repository', function() {
         const patch = await repository.getFilePatchForPath('a.txt');
         await writeFile(path.join(workdir, 'a.txt'), 'foo\nfoo-1\nfoo-2\n');
 
-        await assertCorrectInvalidation({repository, verbose: true}, async () => {
+        await assertCorrectInvalidation({repository}, async () => {
           await repository.applyPatchToIndex(patch);
         });
       });

--- a/test/models/repository.test.js
+++ b/test/models/repository.test.js
@@ -1001,6 +1001,7 @@ describe('Repository', function() {
         op: () => repository.getBranches(),
       });
       calls.set('getCurrentBranch', {
+        serial: true,
         op: () => repository.getCurrentBranch(),
       });
       calls.set('getRemotes', {

--- a/test/models/repository.test.js
+++ b/test/models/repository.test.js
@@ -1062,10 +1062,10 @@ describe('Repository', function() {
         const results = new Map();
 
         for (const [name, call] of methods) {
-          const promise = call.op();
+          const promise = call();
           results.set(name, promise);
           if (process.platform === 'win32') {
-            await promise;
+            await promise.catch(() => {});
           }
         }
 

--- a/test/models/repository.test.js
+++ b/test/models/repository.test.js
@@ -1402,12 +1402,16 @@ describe('Repository', function() {
       }
 
       function expectEvents(repository, ...fileNames) {
-        const pending = new Set(fileNames);
+        const pending = new Set(
+          fileNames.map(fileName => {
+            return path.join(repository.getWorkingDirectoryPath(), ...fileName.split(/[\\/]/));
+          }),
+        );
         return new Promise((resolve, reject) => {
           eventCallback = () => {
             const matchingPaths = observedEvents
               .map(event => event.path)
-              .filter(eventPath => pending.delete(path.basename(eventPath)));
+              .filter(eventPath => pending.delete(eventPath));
 
             if (matchingPaths.length > 0) {
               repository.observeFilesystemChange(matchingPaths);
@@ -1436,7 +1440,7 @@ describe('Repository', function() {
         await assertCorrectInvalidation({repository}, async () => {
           await observer.start();
           await repository.git.stageFiles(['a.txt']);
-          await expectEvents(repository, 'index');
+          await expectEvents(repository, '.git/index');
         });
       });
 
@@ -1449,7 +1453,7 @@ describe('Repository', function() {
         await assertCorrectInvalidation({repository}, async () => {
           await observer.start();
           await repository.git.unstageFiles(['a.txt']);
-          await expectEvents(repository, 'index');
+          await expectEvents(repository, '.git/index');
         });
       });
 
@@ -1459,7 +1463,7 @@ describe('Repository', function() {
         await assertCorrectInvalidation({repository}, async () => {
           await observer.start();
           await repository.git.unstageFiles(['a.txt'], 'HEAD~');
-          await expectEvents(repository, 'index');
+          await expectEvents(repository, '.git/index');
         });
       });
 
@@ -1472,7 +1476,7 @@ describe('Repository', function() {
         await assertCorrectInvalidation({repository}, async () => {
           await observer.start();
           await repository.git.applyPatch(patch.getHeaderString() + patch.toString(), {index: true});
-          await expectEvents(repository, 'index');
+          await expectEvents(repository, '.git/index');
         });
       });
 
@@ -1498,7 +1502,7 @@ describe('Repository', function() {
         await assertCorrectInvalidation({repository}, async () => {
           await observer.start();
           await repository.git.commit('boop your snoot');
-          await expectEvents(repository, 'index', 'master');
+          await expectEvents(repository, '.git/index', '.git/refs/heads/master');
         });
       });
 
@@ -1508,7 +1512,7 @@ describe('Repository', function() {
         await assertCorrectInvalidation({repository}, async () => {
           await observer.start();
           await assert.isRejected(repository.git.merge('origin/branch'));
-          await expectEvents(repository, 'index', 'modified-on-both-ours.txt', 'MERGE_HEAD');
+          await expectEvents(repository, '.git/index', 'modified-on-both-ours.txt', '.git/MERGE_HEAD');
         });
       });
 
@@ -1519,7 +1523,7 @@ describe('Repository', function() {
         await assertCorrectInvalidation({repository}, async () => {
           await observer.start();
           await repository.git.abortMerge();
-          await expectEvents(repository, 'index', 'modified-on-both-ours.txt', 'MERGE_HEAD', 'HEAD');
+          await expectEvents(repository, '.git/index', 'modified-on-both-ours.txt', '.git/MERGE_HEAD');
         });
       });
 
@@ -1529,7 +1533,7 @@ describe('Repository', function() {
         await assertCorrectInvalidation({repository}, async () => {
           await observer.start();
           await repository.git.checkout('HEAD^');
-          await expectEvents(repository, 'index', 'HEAD', 'b.txt', 'c.txt');
+          await expectEvents(repository, '.git/index', '.git/HEAD', 'b.txt', 'c.txt');
         });
       });
 
@@ -1539,7 +1543,7 @@ describe('Repository', function() {
         await assertCorrectInvalidation({repository}, async () => {
           await observer.start();
           await repository.git.checkoutFiles(['b.txt'], 'HEAD^');
-          await expectEvents(repository, 'b.txt', 'index');
+          await expectEvents(repository, 'b.txt', '.git/index');
         });
       });
 
@@ -1553,7 +1557,7 @@ describe('Repository', function() {
         await assertCorrectInvalidation({repository}, async () => {
           await observer.start();
           await repository.git.fetch('origin', 'master');
-          await expectEvents(repository, 'master');
+          await expectEvents(repository, '.git/refs/remotes/origin/master');
         });
       });
 
@@ -1568,7 +1572,7 @@ describe('Repository', function() {
         await assertCorrectInvalidation({repository}, async () => {
           await observer.start();
           await assert.isRejected(repository.git.pull('origin', 'master'));
-          await expectEvents(repository, 'file.txt', 'master', 'MERGE_HEAD', 'index');
+          await expectEvents(repository, 'file.txt', '.git/refs/remotes/origin/master', '.git/MERGE_HEAD', '.git/index');
         });
       });
 
@@ -1583,7 +1587,7 @@ describe('Repository', function() {
         await assertCorrectInvalidation({repository}, async () => {
           await observer.start();
           await repository.git.push('origin', 'master');
-          await expectEvents(repository, 'master');
+          await expectEvents(repository, '.git/refs/remotes/origin/master');
         });
       });
 
@@ -1594,7 +1598,7 @@ describe('Repository', function() {
         await assertCorrectInvalidation({repository, optionNames}, async () => {
           await observer.start();
           await repository.git.setConfig('core.editor', 'ed # :trollface:');
-          await expectEvents(repository, 'config');
+          await expectEvents(repository, '.git/config');
         });
       });
 

--- a/test/models/repository.test.js
+++ b/test/models/repository.test.js
@@ -984,27 +984,39 @@ describe('Repository', function() {
       const repository = options.repository;
       const calls = new Map();
 
-      calls.set('getStatusesForChangedFiles', () => repository.getStatusesForChangedFiles());
-      calls.set('getStagedChangesSinceParentCommit', () => repository.getStagedChangesSinceParentCommit());
-      calls.set('getLastCommit', () => repository.getLastCommit());
-      calls.set('getBranches', () => repository.getBranches());
-      calls.set('getCurrentBranch', () => repository.getCurrentBranch());
-      calls.set('getRemotes', () => repository.getRemotes());
+      calls.set('getStatusesForChangedFiles', {
+        serial: true,
+        op: () => repository.getStatusesForChangedFiles(),
+      });
+      calls.set('getStagedChangesSinceParentCommit', {
+        op: () => repository.getStagedChangesSinceParentCommit(),
+      });
+      calls.set('getLastCommit', {
+        op: () => repository.getLastCommit(),
+      });
+      calls.set('getBranches', {
+        op: () => repository.getBranches(),
+      });
+      calls.set('getCurrentBranch', {
+        op: () => repository.getCurrentBranch(),
+      });
+      calls.set('getRemotes', {
+        op: () => repository.getRemotes(),
+      });
 
       const withFile = fileName => {
-        calls.set(
-          `getFilePatchForPath {unstaged} ${fileName}`,
-          () => repository.getFilePatchForPath(fileName, {staged: false}),
-        );
-        calls.set(
-          `getFilePatchForPath {staged} ${fileName}`,
-          () => repository.getFilePatchForPath(fileName, {staged: true}),
-        );
-        calls.set(
-          `getFilePatchForPath {staged, amending} ${fileName}`,
-          () => repository.getFilePatchForPath(fileName, {staged: true, amending: true}),
-        );
-        calls.set(`readFileFromIndex ${fileName}`, () => repository.readFileFromIndex(fileName));
+        calls.set(`getFilePatchForPath {unstaged} ${fileName}`, {
+          op: () => repository.getFilePatchForPath(fileName, {staged: false}),
+        });
+        calls.set(`getFilePatchForPath {staged} ${fileName}`, {
+          op: () => repository.getFilePatchForPath(fileName, {staged: true}),
+        });
+        calls.set(`getFilePatchForPath {staged, amending} ${fileName}`, {
+          op: () => repository.getFilePatchForPath(fileName, {staged: true, amending: true}),
+        });
+        calls.set(`readFileFromIndex ${fileName}`, {
+          op: () => repository.readFileFromIndex(fileName),
+        });
       };
 
       for (const fileName of await filesWithinRepository(options.repository)) {
@@ -1012,16 +1024,26 @@ describe('Repository', function() {
       }
 
       const withBranch = (branchName, description) => {
-        calls.set(`getAheadCount ${description}`, () => repository.getAheadCount(branchName));
-        calls.set(`getBehindCount ${description}`, () => repository.getBehindCount(branchName));
+        calls.set(`getAheadCount ${description}`, {
+          serial: true,
+          op: () => repository.getAheadCount(branchName),
+        });
+        calls.set(`getBehindCount ${description}`, {
+          serial: true,
+          op: () => repository.getBehindCount(branchName),
+        });
       };
       for (const branchName of await repository.git.getBranches()) {
         withBranch(branchName);
       }
 
       for (const optionName of (options.optionNames || [])) {
-        calls.set(`getConfig ${optionName}`, () => repository.getConfig(optionName));
-        calls.set(`getConfig {local} ${optionName}`, () => repository.getConfig(optionName, {local: true}));
+        calls.set(`getConfig ${optionName}`, {
+          op: () => repository.getConfig(optionName),
+        });
+        calls.set(`getConfig {local} ${optionName}`, {
+          op: () => repository.getConfig(optionName, {local: true}),
+        });
       }
 
       return calls;
@@ -1036,11 +1058,29 @@ describe('Repository', function() {
         methods.delete(opName);
       }
 
-      const record = () => {
+      const record = async () => {
         const results = new Map();
+
+        const serial = [];
+        const parallel = [];
         for (const [name, call] of methods) {
-          results.set(name, call());
+          if (call.serial) {
+            serial.push([name, call]);
+          } else {
+            parallel.push([name, call]);
+          }
         }
+
+        for (const [name, call] of serial) {
+          const promise = call.op();
+          results.set(name, promise);
+          await promise;
+        }
+
+        for (const [name, call] of parallel) {
+          results.set(name, call.op());
+        }
+
         return results;
       };
 
@@ -1074,9 +1114,9 @@ describe('Repository', function() {
         );
       };
 
-      const before = record();
+      const before = await record();
       await operation();
-      const cached = record();
+      const cached = await record();
 
       options.repository.state.cache.clear();
       const after = await record();

--- a/test/models/repository.test.js
+++ b/test/models/repository.test.js
@@ -987,67 +987,63 @@ describe('Repository', function() {
       const repository = options.repository;
       const calls = new Map();
 
-      calls.set('getStatusesForChangedFiles', {
-        serial: true,
-        op: () => repository.getStatusesForChangedFiles(),
-      });
-      calls.set('getStagedChangesSinceParentCommit', {
-        op: () => repository.getStagedChangesSinceParentCommit(),
-      });
-      calls.set('getLastCommit', {
-        op: () => repository.getLastCommit(),
-      });
-      calls.set('getBranches', {
-        op: () => repository.getBranches(),
-      });
-      calls.set('getCurrentBranch', {
-        serial: true,
-        op: () => repository.getCurrentBranch(),
-      });
-      calls.set('getRemotes', {
-        op: () => repository.getRemotes(),
-      });
+      calls.set(
+        'getStatusBundle',
+        () => repository.getStatusBundle(),
+      );
+      calls.set(
+        'getHeadDescription',
+        () => repository.getHeadDescription(),
+      );
+      calls.set(
+        'getStagedChangesSinceParentCommit',
+        () => repository.getStagedChangesSinceParentCommit(),
+      );
+      calls.set(
+        'getLastCommit',
+        () => repository.getLastCommit(),
+      );
+      calls.set(
+        'getBranches',
+        () => repository.getBranches(),
+      );
+      calls.set(
+        'getRemotes',
+        () => repository.getRemotes(),
+      );
 
       const withFile = fileName => {
-        calls.set(`getFilePatchForPath {unstaged} ${fileName}`, {
-          op: () => repository.getFilePatchForPath(fileName, {staged: false}),
-        });
-        calls.set(`getFilePatchForPath {staged} ${fileName}`, {
-          op: () => repository.getFilePatchForPath(fileName, {staged: true}),
-        });
-        calls.set(`getFilePatchForPath {staged, amending} ${fileName}`, {
-          op: () => repository.getFilePatchForPath(fileName, {staged: true, amending: true}),
-        });
-        calls.set(`readFileFromIndex ${fileName}`, {
-          op: () => repository.readFileFromIndex(fileName),
-        });
+        calls.set(
+          `getFilePatchForPath {unstaged} ${fileName}`,
+          () => repository.getFilePatchForPath(fileName, {staged: false}),
+        );
+        calls.set(
+          `getFilePatchForPath {staged} ${fileName}`,
+          () => repository.getFilePatchForPath(fileName, {staged: true}),
+        );
+        calls.set(
+          `getFilePatchForPath {staged, amending} ${fileName}`,
+          () => repository.getFilePatchForPath(fileName, {staged: true, amending: true}),
+        );
+        calls.set(
+          `readFileFromIndex ${fileName}`,
+          () => repository.readFileFromIndex(fileName),
+        );
       };
 
       for (const fileName of await filesWithinRepository(options.repository)) {
         withFile(fileName);
       }
 
-      const withBranch = branchName => {
-        calls.set(`getAheadCount ${branchName}`, {
-          serial: true,
-          op: () => repository.getAheadCount(branchName),
-        });
-        calls.set(`getBehindCount ${branchName}`, {
-          serial: true,
-          op: () => repository.getBehindCount(branchName),
-        });
-      };
-      for (const branchName of await repository.git.getBranches()) {
-        withBranch(branchName);
-      }
-
       for (const optionName of (options.optionNames || [])) {
-        calls.set(`getConfig ${optionName}`, {
-          op: () => repository.getConfig(optionName),
-        });
-        calls.set(`getConfig {local} ${optionName}`, {
-          op: () => repository.getConfig(optionName, {local: true}),
-        });
+        calls.set(
+          `getConfig ${optionName}`,
+          () => repository.getConfig(optionName),
+        );
+        calls.set(
+          `getConfig {local} ${optionName}`,
+          () => repository.getConfig(optionName, {local: true}),
+        );
       }
 
       return calls;
@@ -1065,24 +1061,12 @@ describe('Repository', function() {
       const record = async () => {
         const results = new Map();
 
-        const serial = [];
-        const parallel = [];
         for (const [name, call] of methods) {
-          if (call.serial) {
-            serial.push([name, call]);
-          } else {
-            parallel.push([name, call]);
-          }
-        }
-
-        for (const [name, call] of serial) {
           const promise = call.op();
           results.set(name, promise);
-          await promise;
-        }
-
-        for (const [name, call] of parallel) {
-          results.set(name, call.op());
+          if (process.platform === 'win32') {
+            await promise;
+          }
         }
 
         return results;

--- a/test/models/repository.test.js
+++ b/test/models/repository.test.js
@@ -1404,7 +1404,7 @@ describe('Repository', function() {
       function expectEvents(repository, ...fileNames) {
         const pending = new Set(
           fileNames.map(fileName => {
-            return fs.realPathSync(
+            return fs.realpathSync(
               path.join(repository.getWorkingDirectoryPath(), ...fileName.split(/[\\/]/)),
             );
           }),
@@ -1412,7 +1412,7 @@ describe('Repository', function() {
         return new Promise((resolve, reject) => {
           eventCallback = () => {
             const matchingPaths = observedEvents
-              .map(event => fs.realPathSync(event.path))
+              .map(event => fs.realpathSync(event.path))
               .filter(eventPath => pending.delete(eventPath));
 
             if (matchingPaths.length > 0) {

--- a/test/models/repository.test.js
+++ b/test/models/repository.test.js
@@ -1404,13 +1404,15 @@ describe('Repository', function() {
       function expectEvents(repository, ...fileNames) {
         const pending = new Set(
           fileNames.map(fileName => {
-            return path.join(repository.getWorkingDirectoryPath(), ...fileName.split(/[\\/]/));
+            return fs.realPathSync(
+              path.join(repository.getWorkingDirectoryPath(), ...fileName.split(/[\\/]/)),
+            );
           }),
         );
         return new Promise((resolve, reject) => {
           eventCallback = () => {
             const matchingPaths = observedEvents
-              .map(event => event.path)
+              .map(event => fs.realPathSync(event.path))
               .filter(eventPath => pending.delete(eventPath));
 
             if (matchingPaths.length > 0) {

--- a/test/models/repository.test.js
+++ b/test/models/repository.test.js
@@ -1429,7 +1429,7 @@ describe('Repository', function() {
         await assertCorrectInvalidation({repository}, async () => {
           await observer.start();
           await repository.git.stageFiles(['a.txt']);
-          await expectEvents(repository, '.git/index');
+          await expectEvents(repository, path.join('.git', 'index'));
         });
       });
 
@@ -1442,7 +1442,7 @@ describe('Repository', function() {
         await assertCorrectInvalidation({repository}, async () => {
           await observer.start();
           await repository.git.unstageFiles(['a.txt']);
-          await expectEvents(repository, '.git/index');
+          await expectEvents(repository, path.join('.git', 'index'));
         });
       });
 
@@ -1452,7 +1452,7 @@ describe('Repository', function() {
         await assertCorrectInvalidation({repository}, async () => {
           await observer.start();
           await repository.git.unstageFiles(['a.txt'], 'HEAD~');
-          await expectEvents(repository, '.git/index');
+          await expectEvents(repository, path.join('.git', 'index'));
         });
       });
 
@@ -1465,7 +1465,10 @@ describe('Repository', function() {
         await assertCorrectInvalidation({repository}, async () => {
           await observer.start();
           await repository.git.applyPatch(patch.getHeaderString() + patch.toString(), {index: true});
-          await expectEvents(repository, '.git/index');
+          await expectEvents(
+            repository,
+            path.join('.git', 'index'),
+          );
         });
       });
 
@@ -1478,7 +1481,10 @@ describe('Repository', function() {
         await assertCorrectInvalidation({repository}, async () => {
           await observer.start();
           await repository.git.applyPatch(patch.getHeaderString() + patch.toString());
-          await expectEvents(repository, 'a.txt');
+          await expectEvents(
+            repository,
+            'a.txt',
+          );
         });
       });
 
@@ -1491,7 +1497,11 @@ describe('Repository', function() {
         await assertCorrectInvalidation({repository}, async () => {
           await observer.start();
           await repository.git.commit('boop your snoot');
-          await expectEvents(repository, '.git/index', '.git/refs/heads/master');
+          await expectEvents(
+            repository,
+            path.join('.git', 'index'),
+            path.join('.git', 'refs', 'heads', 'master'),
+          );
         });
       });
 
@@ -1501,7 +1511,12 @@ describe('Repository', function() {
         await assertCorrectInvalidation({repository}, async () => {
           await observer.start();
           await assert.isRejected(repository.git.merge('origin/branch'));
-          await expectEvents(repository, '.git/index', 'modified-on-both-ours.txt', '.git/MERGE_HEAD');
+          await expectEvents(
+            repository,
+            path.join('.git', 'index'),
+            'modified-on-both-ours.txt',
+            path.join('.git', 'MERGE_HEAD'),
+          );
         });
       });
 
@@ -1512,7 +1527,12 @@ describe('Repository', function() {
         await assertCorrectInvalidation({repository}, async () => {
           await observer.start();
           await repository.git.abortMerge();
-          await expectEvents(repository, '.git/index', 'modified-on-both-ours.txt', '.git/MERGE_HEAD');
+          await expectEvents(
+            repository,
+            path.join('.git', 'index'),
+            'modified-on-both-ours.txt',
+            path.join('.git', 'MERGE_HEAD'),
+          );
         });
       });
 
@@ -1522,7 +1542,13 @@ describe('Repository', function() {
         await assertCorrectInvalidation({repository}, async () => {
           await observer.start();
           await repository.git.checkout('HEAD^');
-          await expectEvents(repository, '.git/index', '.git/HEAD', 'b.txt', 'c.txt');
+          await expectEvents(
+            repository,
+            path.join('.git', 'index'),
+            path.join('.git', 'HEAD'),
+            'b.txt',
+            'c.txt',
+          );
         });
       });
 
@@ -1532,7 +1558,11 @@ describe('Repository', function() {
         await assertCorrectInvalidation({repository}, async () => {
           await observer.start();
           await repository.git.checkoutFiles(['b.txt'], 'HEAD^');
-          await expectEvents(repository, 'b.txt', '.git/index');
+          await expectEvents(
+            repository,
+            'b.txt',
+            path.join('.git', 'index'),
+          );
         });
       });
 
@@ -1546,7 +1576,10 @@ describe('Repository', function() {
         await assertCorrectInvalidation({repository}, async () => {
           await observer.start();
           await repository.git.fetch('origin', 'master');
-          await expectEvents(repository, '.git/refs/remotes/origin/master');
+          await expectEvents(
+            repository,
+            path.join('.git', 'refs', 'remotes', 'origin', 'master'),
+          );
         });
       });
 
@@ -1561,7 +1594,13 @@ describe('Repository', function() {
         await assertCorrectInvalidation({repository}, async () => {
           await observer.start();
           await assert.isRejected(repository.git.pull('origin', 'master'));
-          await expectEvents(repository, 'file.txt', '.git/refs/remotes/origin/master', '.git/MERGE_HEAD', '.git/index');
+          await expectEvents(
+            repository,
+            'file.txt',
+            path.join('.git', 'refs', 'remotes', 'origin', 'master'),
+            path.join('.git', 'MERGE_HEAD'),
+            path.join('.git', 'index'),
+          );
         });
       });
 
@@ -1576,7 +1615,10 @@ describe('Repository', function() {
         await assertCorrectInvalidation({repository}, async () => {
           await observer.start();
           await repository.git.push('origin', 'master');
-          await expectEvents(repository, '.git/refs/remotes/origin/master');
+          await expectEvents(
+            repository,
+            path.join('.git', 'refs', 'remotes', 'origin', 'master'),
+          );
         });
       });
 
@@ -1587,7 +1629,10 @@ describe('Repository', function() {
         await assertCorrectInvalidation({repository, optionNames}, async () => {
           await observer.start();
           await repository.git.setConfig('core.editor', 'ed # :trollface:');
-          await expectEvents(repository, '.git/config');
+          await expectEvents(
+            repository,
+            path.join('.git', 'config'),
+          );
         });
       });
 
@@ -1597,7 +1642,10 @@ describe('Repository', function() {
         await assertCorrectInvalidation({repository}, async () => {
           await observer.start();
           await writeFile(path.join(workdir, 'b.txt'), 'new contents\n');
-          await expectEvents(repository, 'b.txt');
+          await expectEvents(
+            repository,
+            'b.txt',
+          );
         });
       });
     });

